feat: add support for read in edge environments

vnphanquang opened the pull request:
closes https://github.com/sveltejs/kit/issues/12446
closes https://github.com/sveltejs/kit/pull/11674

This PR helps with the cloudflare part of #12446. Added a test case and I also tested in a project bootstrapped with `create cloudflare@latest`. Hope i didn't miss anything 🤞.

Thanks all.

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

vnphanquang commented:
Not sure why that particular firefox CI test failed but it did pass on my local machine. Let me know, thanks!

![image](https://github.com/user-attachments/assets/9c1be135-0430-4d9e-b46f-22010e120a99)

eltigerchino commented on packages/adapter-cloudflare/test/apps/goose.jpg:
We could save some space by testing with a .txt file instead. The .txt file has to be imported with the `?url` query so that Vite doesn't inline it. For example:
```js
import textFile from './myFile.txt?url';
```

eltigerchino replied:
Oh I'm not actually sure. I would think that it doesn't but I'd have to check if that's something Cloudflare is doing or if it happens in Node as well.

vnphanquang replied:
Just gave it a quick test with adapter-node. `content-type` is correctly returned from the built server, but not during development. Here's a [test repo](https://github.com/vnphanquang/adapter-node-read-test/blob/master/src/routes/read/%2Bserver.ts) if you want to take a look.

Anyway, let's simplify to `txt` for now. Having correct `mime` is nice but it seems out of scope for this particular PR.

eltigerchino replied:
It's probably because during development, `read` uses a different implementation (it assumes we're always running in a Node.js environment).
https://github.com/sveltejs/kit/blob/6261a877ae97ff85d07277c51391d925ed1bd096/packages/kit/src/exports/vite/dev/index.js#L550-L556
https://github.com/sveltejs/kit/blob/6261a877ae97ff85d07277c51391d925ed1bd096/packages/kit/src/exports/vite/preview/index.js#L203-L209
This could be an issue once dev runs things in a Cloudflare or Netlify environment etc. based on the adapter.

eltigerchino commented:
Thanks for this PR! Is there any chance you could continue the work from https://github.com/sveltejs/kit/pull/11674 so that it's implemented in the Netlify and Vercel adapters too? I think if the reusable `fetchFile` function could allow passing a `fetch` function to it, we could pass `env.ASSETS.fetch` in the case of the Cloudflare adapter and it would just work.

> Not sure why that particular firefox CI test failed but it did pass on my local machine (`pnpm test:cross-platform:build`). Let me know, thanks!

I re-ran the test in CI and it passes. Probably just a flaky test that failed.

vnphanquang commented:
Ah yes. Sorry I missed Rich's PR in my search. Will look into it. Thanks for the input!

vladshcherbin commented:
Can this be prioritized? It's very cumbersome to use static assets with Cloudflare with unexpected outcome. Vercel and other can be added later

eltigerchino commented:
I've added support to the Netlify and Vercel adapters. Also added a test suite for Netlify now that the local dev environment supports edge functions.

vnphanquang commented:
> I've added support to the Netlify and Vercel adapters. Also added a test suite for Netlify now that the local dev environment supports edge functions.

Thanks @eltigerchino. Apologies I couldn't find time to follow through with Netlify & Vercel. Appreciate you taking on the responsibility!

dominikg commented on packages/adapter-vercel/package.json:
```diff
@@ -28,29 +28,35 @@
 	],
 	"scripts": {
+		"dev": "node -e \"fs.rmSync('files', { force: true, recursive: true })\" && rollup -cw",
+		"build": "node -e \"fs.rmSync('files', { force: true, recursive: true })\" && rollup -c && node -e \"fs.cpSync('src/serverless.js', 'files/serverless.js');\"",
```
see above.

eltigerchino commented on packages/adapter-netlify/test/apps/edge/deno.lock:
```diff
@@ -0,0 +1,19 @@
+      ]
+    }
+  }
+}
```
This is a build artifact from running the app with the Netlify CLI. I think it those another bundle at that point. I'll add it to to the .gitignore

dummdidumm commented:
I think eventually we wanna turn these into "deploy to prod, do playwright stuff against prod url" tests. Slower but more reliable and fully replicates the environment

benmccann commented:
In the meantime I wonder if we could just invoke `sirv` since the only thing we do is call `netlify serve`. I'm not sure what all `netlify serve` does and it's probably not exactly the same, but it would be a lot lighter.

eltigerchino commented:
> In the meantime I wonder if we could just invoke `sirv` since the only thing we do is call `netlify serve`. I'm not sure what all `netlify serve` does and it's probably not exactly the same, but it would be a lot lighter.

I'm not sure how sirv would work with the Netlify build output. At least in https://github.com/sveltejs/kit/pull/12296 it's important for testing the split functions functionality with the Netlify middleware.

The Environment API integration is still far off, otherwise we could use the Netlify Vite plugin instead. I think Simon's idea might be a lot better and something we can implement sooner.

cc: @serhalp

benmccann commented:
I shared another idea here: https://github.com/e18e/ecosystem-issues/issues/156#issuecomment-3160886873
